python/cudf/cudf/testing/testing.py

```diff
@@ -0,0 +1,581 @@
+        Not yet supported
+    check_exact : bool, default False
+        Whether to compare number exactly.
+    check_datetime_like_compat : bool, default False
+        Compare datetime-like which is comparable ignoring dtype.
+    check_categorical : bool, default True
+        Whether to compare internal Categorical exactly.
+    check_category_order : bool, default True
+        Whether to compare category order of internal Categoricals
+    obj : str, default ‘ColumnBase’
+        Specify object name being compared, internally used to
+        show appropriate assertion message.
+    """
+    if check_dtype:
+        if (
+            is_categorical_dtype(left)
+            and is_categorical_dtype(right)
+            and not check_categorical
+        ):
+            pass
+        else:
+            if type(left) != type(right) or left.dtype != right.dtype:
+                msg1 = f"{left.dtype}"
+                msg2 = f"{right.dtype}"
+                raise_assert_detail(obj, "Dtypes are different", msg1, msg2)
+
+    if check_datetimelike_compat:
+        if np.issubdtype(left.dtype, np.datetime64):
+            right = right.astype(left.dtype)
+        elif np.issubdtype(right.dtype, np.datetime64):
+            left = left.astype(right.dtype)
+
+        if np.issubdtype(left.dtype, np.datetime64):
+            if not left.equals(right):
+                raise AssertionError(
+                    f"[datetimelike_compat=True] {left.values} "
+                    f"is not equal to {right.values}."
+                )
+            return
+
+    if check_exact and check_categorical:
+        if is_categorical_dtype(left) and is_categorical_dtype(right):
+            left_cat = left.cat().categories
+            right_cat = right.cat().categories
+
+            if check_category_order:
+                assert_index_equal(
+                    left_cat,
+                    right_cat,
+                    exact=check_dtype,
+                    check_exact=True,
+                    check_categorical=False,
+                )
+                assert_column_equal(
+                    left.codes,
+                    right.codes,
+                    check_dtype=check_dtype,
+                    check_exact=True,
+                    check_categorical=False,
+                    check_category_order=False,
+                )
+
+            if left.ordered != right.ordered:
+                msg1 = f"{left.ordered}"
+                msg2 = f"{right.ordered}"
+                raise_assert_detail(
+                    "{obj} category", "Orders are different", msg1, msg2
+                )
+
+    if (
+        not check_dtype
+        and is_categorical_dtype(left)
+        and is_categorical_dtype(right)
+    ):
+        left = left.astype(left.categories.dtype)
+        right = right.astype(right.categories.dtype)
+
+    columns_equal = False
+    try:
+        columns_equal = left.equals(right)
+    except TypeError:
```

Comment:
Can we document what we're expecting to raise this `TypeError` as it's not clear

Reply:
When we compare to two categorical columns, and if the categories don't match it will throw TypeError, I will also update it with a comment.

Reply:
OK. The logic is a bit hard to follow here.

```
    try:
        columns_equal = left.equals(right)
    except TypeError:
        if is_categorical_dtype(left) and is_categorical_dtype(right):
            left = left.astype(left.categories.dtype)
            right = right.astype(right.categories.dtype)
```

What if a `TypeError` is raised and `left` and `right` are *not* categoricals? Then we're effectively ignoring that `TypeError`. Is that the desired behaviour? Or should we re-raise?

Reply:
We should re-raise, I will make sure this is taken care only in the case I have described.